### PR TITLE
Issue #1053: Consolidate URL to website mapping

### LINKED ISSUE #1053: Consolidate URL to Website mapping configuration
Currently the information has to be specified as part of the `SCRIPT_NAME` server environment by the httpd configuration, e.g.

```
location / {
    try_files $uri @lizards_and_pumpkins;
    set $LIZARDS_SCRIPT_NAME /en/index.php;
  }

  location /en/ {
    try_files $uri @lizards_and_pumpkins;
    fastcgi_param SCRIPT_NAME /en/index.php;
  }

  location /de/ {
    try_files $uri @lizards_and_pumpkins;
    fastcgi_param SCRIPT_NAME /de/index.php;
  }

  location /fr/ {
    try_files $uri @lizards_and_pumpkins;
    fastcgi_param SCRIPT_NAME /fr/index.php;
  }
```

In addition to this, the request is also mapped to a website in the `LP_BASE_URL_TO_WEBSITE_MAP` environment variable, e.g.

```
LP_BASE_URL_TO_WEBSITE_MAP "http://demo.lizardsandpumpkins.com.loc/en/=en|http://demo.lizardsandpumpkins.com.loc/de/=de|http://demo.lizardsandpumpkins.com.loc/fr/=fr|http://demo.lizardsandpumpkins.com.loc/=en";
```

This is duplicate knowledge which makes future changes harder, introduces the risk of inconsistencies, and more work when new websites are being introduced.

Instead of the method `\LizardsAndPumpkins\Http\HttpUrl::getPathWithoutWebsitePrefix()` using the `SCRIPT_NAME` value to determine the path without the website prefix, the interface `\LizardsAndPumpkins\Context\Website\UrlToWebsiteMap` should get a new method to do so, e.g. `getRequestPathWithoutWebsitePrefix(string $url)`.

After the new method is introduced, all usages of `HttpUrl::getPathWithoutWebsitePrefix()` in the project should be refactored to use the new method.  
Then `HttpUrl::getPathWithoutWebsitePrefix()` should be `@deprecated`.

### PR DESCRIPTION
**Pull request by Vinai**
Please refer to #1053 for details on the issue.

The setting of the `SCRIPT_NAME` server variable in the nginx configuration is no longer required with these code changes, as the request path without website prefix is determined based on the `LP_BASE_URL_TO_WEBSITE_MAP` [environment variable](https://github.com/lizards-and-pumpkins/dev-vm/blob/master/provisioning/sample/demo.lizardsandpumpkins.com.loc#L63).

Closes #1053

### CODE REVIEW

File: src/Context/Website/ConfigurableUrlToWebsiteMap.php [hunk 1]
```diff
@@ -75,25 +75,38 @@ private static function splitConfigRecord(string $mapping) : array
...
-     * @return string[]
-     */
-    private static function flatten(array $array) : array
+    private function getWebsiteUrlPrefixAndCodeByUrl(string $url): array
     {
-        return array_reduce($array, function (array $carry, array $pair) {
-            return array_merge($carry, $pair);
-        }, []);
+        foreach ($this->urlToWebsiteMap as $urlPrefix => $website) {
+            if (stripos($url, $urlPrefix) === 0) {
```
Comment (Schrank) [inline src/Context/Website/ConfigurableUrlToWebsiteMap.php]: Technically domains are case INsenstive, directories are not. 

    example.com/de != example.com/dE != example.com/De != example.com/DE

I guess no one implements it this way and I think we shouldn't either.

File: src/Context/Website/ConfigurableUrlToWebsiteMap.php [hunk 2]
```diff
@@ -75,25 +75,38 @@ private static function splitConfigRecord(string $mapping) : array
...
+        list($urlPrefix, $website) = $this->getWebsiteUrlPrefixAndCodeByUrl($url);
+        $pathWithoutPrefix = substr($url, strlen($urlPrefix));
+
+        return $this->hasQueryParameter($url) ? $this->removeQueryParameter($pathWithoutPrefix) : $pathWithoutPrefix;
+    }
 
-        throw new UnknownWebsiteUrlException(sprintf('No website code found for url "%s"', $url));
+    private function hasQueryParameter(string $url): bool
+    {
+        return strrpos($url, '?') !== false;
```
Comment (Schrank) [inline src/Context/Website/ConfigurableUrlToWebsiteMap.php]: This should give false return on:

    http://example.com/directory/#yoMan?

Better use `parse_url($url, PHP_URL_QUERY) !== ''`

File: src/Context/Website/ConfigurableUrlToWebsiteMap.php [hunk 3]
```diff
@@ -75,25 +75,38 @@ private static function splitConfigRecord(string $mapping) : array
...
+    }
+
+    private function removeQueryParameter($pathWithoutPrefix): string
+    {
+        return substr($pathWithoutPrefix, 0, strrpos($pathWithoutPrefix, '?'));
```
Comment (Schrank) [inline src/Context/Website/ConfigurableUrlToWebsiteMap.php]: Same here, better use `parse_url` and recombine what you want

File: src/Http/HttpUrl.php [hunk 4]
```diff
@@ -84,17 +84,16 @@ public function __toString() : string
         return $schema . '//' . $this->host . $port . $this->path . $query;
     }
 
+    /**
+     * @deprecated Use UrlToWebsiteMap::getRequestPathWithoutWebsitePrefix() instead.
```
Comment (Schrank) [inline src/Http/HttpUrl.php]: <3

File: src/ProductDetail/ProductDetailViewRequestHandler.php [hunk 5]
```diff
@@ -104,7 +112,7 @@ private function loadPageMetaInfoSnippet(HttpRequest $request)
             }
         }
     }
-
+
```
Comment (Schrank) [inline src/ProductDetail/ProductDetailViewRequestHandler.php]: Please make sure your IDE is removing white spaces on empty lines

File: src/ProductRelations/ContentDelivery/ProductRelationsApiV1GetRequestHandler.php [hunk 6]
```diff
@@ -24,30 +25,40 @@ class ProductRelationsApiV1GetRequestHandler extends ApiRequestHandler
...
         $parts = $this->getRequestPathParts($request);
+
         return count($parts) > 4 && 'products' === $parts[1] && 'relations' === $parts[3];
     }
 
-    final protected function getResponse(HttpRequest $request) : HttpResponse
+    final protected function getResponse(HttpRequest $request): HttpResponse
     {
-        if (! $this->canProcess($request)) {
+        if (!$this->canProcess($request)) {
```
Comment (Schrank) [inline src/ProductRelations/ContentDelivery/ProductRelationsApiV1GetRequestHandler.php]: I think we decided to add this space

File: src/RestApi/ApiRouter.php [hunk 7]
```diff
@@ -27,7 +34,7 @@ public function __construct(ApiRequestHandlerLocator $requestHandlerLocator)
      */
     public function route(HttpRequest $request)
     {
-        $urlPath = trim($request->getPathWithoutWebsitePrefix(), '/');
+        $urlPath = trim($this->urlToWebsiteMap->getRequestPathWithoutWebsitePrefix((string) $request->getUrl()), '/');
         $urlToken = explode('/', $urlPath);
 
         if (self::API_URL_PREFIX !== array_shift($urlToken)) {
```
Comment (Schrank) [inline src/RestApi/ApiRouter.php]: Have to stop here, work todo.

File: src/Context/Website/ConfigurableUrlToWebsiteMap.php [hunk 8]
```diff
@@ -58,14 +58,14 @@ private static function buildArrayMapFromString(string $configValue)
...
-        return self::flatten($pairs);
+        return array_merge(...$pairs);
     }
 
     /**
      * @param string $mapping
      * @return string[]
      */
-    private static function splitConfigRecord(string $mapping) : array
+    private static function splitConfigRecord(string $mapping): array
```
Comment (Vinai) [inline src/Context/Website/ConfigurableUrlToWebsiteMap.php]: No space.

File: src/Context/Website/ConfigurableUrlToWebsiteMap.php (hunk 2 above)
Comment (Vinai) [inline src/Context/Website/ConfigurableUrlToWebsiteMap.php]: The browser does not send the hash to the server, but yeah, `parse_url` is indeed better.

File: src/ProductDetail/ProductDetailViewRequestHandler.php (hunk 5 above)
Comment (Vinai) [inline src/ProductDetail/ProductDetailViewRequestHandler.php]: Not part of our coding standard.

File: src/Context/Website/ConfigurableUrlToWebsiteMap.php [hunk 9]
```diff
@@ -58,42 +58,57 @@ private static function buildArrayMapFromString(string $configValue)
...
-            if (stripos($url, $urlPattern) === 0) {
-                return $website;
-            }
-        }
+        list($urlPrefix, $website) = $this->getWebsiteUrlPrefixAndCodeByUrl($url);
+
+        return $website;
+    }
+
+    public function getRequestPathWithoutWebsitePrefix(string $url): string
```
Comment (tim-bezhashvyly) [inline src/Context/Website/ConfigurableUrlToWebsiteMap.php]: Or maybe even `HttpRequest` but then I smell LoD violations.

File: tests/Unit/Suites/Context/Website/ConfigurableUrlToWebsiteMapTest.php [hunk 10]
```diff
@@ -82,6 +82,39 @@ public function websiteMapProvider() : array
             ['http://example.com/=foo|http://example.com/=bar', 'http://example.com/', 'bar'],
             ['http://example.com/=foo|https://example.com/=bar', 'http://example.com/', 'foo'],
             ['http://example.com/foo/=foo|http://example.com/bar/=bar', 'http://example.com/bar/baz', 'bar'],
+            ['http://example.com/aa/=foo|http://example.com/=bar', 'http://example.com/aa/baz', 'foo'],
+            ['http://example.com/aa/=foo|http://example.com/=bar', 'http://example.com/baz', 'bar'],
+            ['http://example.com/=bar|http://example.com/aa/=foo', 'http://example.com/aa/baz', 'bar'],
         ];
     }
+
+    public function testThrowsAnExceptionIfTheWebsiteCanNotBeDetermined()
```
Comment (tim-bezhashvyly) [inline tests/Unit/Suites/Context/Website/ConfigurableUrlToWebsiteMapTest.php]: Isn't it duplicating `testExceptionIsThrownIfGivenUrlMatchesNoneOfWebsites` test?

File: tests/Unit/Suites/Context/Website/ConfigurableUrlToWebsiteMapTest.php [hunk 11]
```diff
@@ -82,6 +82,39 @@ public function websiteMapProvider() : array
...
+    {
+        $testMap = 'http://example.com/aa/=foo|http://example.com/=bar';
+        $this->stubConfigReader->method('get')->with(ConfigurableUrlToWebsiteMap::CONFIG_KEY)->willReturn($testMap);
+
+        $websiteMap = ConfigurableUrlToWebsiteMap::fromConfig($this->stubConfigReader);
+        $this->assertSame('a/b/c', $websiteMap->getRequestPathWithoutWebsitePrefix('http://example.com/aa/a/b/c'));
+        $this->assertSame('foo', $websiteMap->getRequestPathWithoutWebsitePrefix('http://example.com/foo'));
+    }
+
+    public function testRturnsTheRequestPathWithoutUrlPrefixWithoutQueryArguments()
```
Comment (tim-bezhashvyly) [inline tests/Unit/Suites/Context/Website/ConfigurableUrlToWebsiteMapTest.php]: Typo in word "returns"

File: src/Context/Website/ConfigurableUrlToWebsiteMap.php (hunk 9 above)
Comment (Vinai) [inline src/Context/Website/ConfigurableUrlToWebsiteMap.php]: Yes, I considered that, but I want to keep the interface consistent:

```php
interface UrlToWebsiteMap
{
    public function getWebsiteCodeByUrl(string $url): Website;
    
    public function getRequestPathWithoutWebsitePrefix(string $url): string;
}
```

Refactoring both methods would probably be better done in a separate PR.

File: tests/Unit/Suites/Context/Website/ConfigurableUrlToWebsiteMapTest.php (hunk 10 above)
Comment (Vinai) [inline tests/Unit/Suites/Context/Website/ConfigurableUrlToWebsiteMapTest.php]: Yes it does. I needed to add the test to move the validation to the right place in the class. Want me to remove it again?

File: tests/Unit/Suites/Context/Website/ConfigurableUrlToWebsiteMapTest.php (hunk 11 above)
Comment (Vinai) [inline tests/Unit/Suites/Context/Website/ConfigurableUrlToWebsiteMapTest.php]: 👍

**Comment by Vinai** [comment]
Ready for next review.

File: tests/Unit/Suites/Context/Website/ConfigurableUrlToWebsiteMapTest.php (hunk 10 above)
Comment (tim-bezhashvyly) [inline tests/Unit/Suites/Context/Website/ConfigurableUrlToWebsiteMapTest.php]: Yes, please.

Comment (Vinai) [inline tests/Unit/Suites/Context/Website/ConfigurableUrlToWebsiteMapTest.php]: Done.
